lib/daru/core/merge.rb

```diff
@@ -1,205 +1,136 @@
+      def row(lkey, rkey)
+        case
+        when !lkey && !rkey
+          # :nocov:
+          # It's just an impossibility handler, can't be covered :)
+          raise 'Unexpected condition met during merge'
+          # :nocov:
+        when lkey == rkey
+          merge_rows(left.shift, right.shift)
+        when !rkey || lt(lkey, rkey)
+          left_row
+        else # !lkey || lt(rkey, lkey)
+          right_row
         end
-
-        @joined_hash
       end
 
-      def df1_hash
-        @df1_hash ||= MergeHelper.hashify @df1
+      def left_row
+        val = left.shift
+        expand_row(val, left_keys) if keep_left
       end
 
-      def df2_hash
-        @df2_hash ||= MergeHelper.hashify @df2
+      def right_row
+        val = right.shift
+        expand_row(val, right_keys) if keep_right
       end
 
-      def df1_array
-        @df1_array ||= MergeHelper.arrayify_with_sort_keys @df1.size, df1_hash, @on
+      def lt(k1, k2)
+        (k1 <=> k2) == -1
       end
 
-      def df2_array
-        @df2_array ||= MergeHelper.arrayify_with_sort_keys @df2.size, df2_hash, @on
+      def merge_rows lrow, rrow
+        left_keys
+          .map { |from, to| [to, lrow[from]] }.to_h
+          .merge(on.map { |col| [col, lrow[col]] }.to_h)
+          .merge(right_keys.map { |from, to| [to, rrow[from]] }.to_h)
       end
 
-      def df1_keys
-        df1_hash.keys
+      def expand_row row, renamings
+        renamings
+          .map { |from, to| [to, row[from]] }.to_h
+          .merge(on.map { |col| [col, row[col]] }.to_h)
       end
 
-      def df2_keys
-        df2_hash.keys
+      def first_key(arr)
+        arr.empty? ? nil : arr.first.values_at(*on)
       end
 
-      # Private: The merge row contains two elements, the first is the row from the
-      # first dataframe, the second is the row from the second dataframe.
-      def add_merge_row_to_hash row, hash
-        @df1_key_to_index ||= df1_keys.each_with_index.map { |k,idx| [k, idx] }.to_h
-        @df2_key_to_index ||= df2_keys.each_with_index.map { |k,idx| [k, idx] }.to_h
-
-        hash.each do |k,v|
-          v ||= []
-
-          left  = df1_keys.include?(k) ? row[0] && row[0][1][@df1_key_to_index[k]] : nil
-          right = df2_keys.include?(k) ? row[1] && row[1][1][@df2_key_to_index[k]] : nil
-
-          v << (left || right)
+      def validate_on!(left_df, right_df)
+        @on.each do |on|
+          left_df.has_vector?(on) && right_df.has_vector?(on) or
+            raise ArgumentError, "Both dataframes expected to have #{on.inspect} field"
         end
       end
     end
 
-    # Private module containing methods for join, merge, concat operations on
-    # dataframes and vectors.
-    # @private
     module Merge
       class << self
         def join df1, df2, opts={}
-          on = opts[:on]
-
-          mf = MergeFrame.new df1, df2, on: on
-          mf.send opts[:how], {}
+          MergeFrame.new(df1, df2, opts).join
```

Comment:
Wow that's some great refactoring. Are you sure it's as fast/faster than the previous implementation? Also, what if we keep the profile directory instead of gitignoring it?